File: ios/LiveTales/uoat/Book.swift
//
//  Book.swift
//  uoat
//
//  Created by Pyro User on 4/5/16.
//  Copyright © 2016 Zed. All rights reserved.
//

import Foundation

class Book
{
    init( id:String, langId:String, isHD:Bool )
    {
        self.id = id
        self.langId = langId
        self.isHD = isHD
        self.story = [String:AnyObject]()
        self.storyLanguage = [String:String]()
        self.imagesPath = [String:String]()
        self.currentPieceStory = 0
        self.currentPage = 0
        self.state = .FrontCover
    }
    
    func getId() -> String
    {
        return self.id
    }
    
    func load() -> Page
    {
        self.loadStory()
        self.loadStoryLanguage()
        self.loadImages()
        
        self.state = .FrontCover

        let pageToShow:Page = self.preparePageToShow( true )
        return pageToShow
    }
    
    func changeLanguage( langId: String ) -> Page
    {
        self.langId = langId
        self.loadStoryLanguage()
        
        let pageToShow:Page = self.preparePageToShow( true )
        return pageToShow
    }
    
    func goBegin() -> Page
    {
        self.currentPieceStory = 0
        self.currentPage = 0
        self.state = .FrontCover
        
        let pageToShow:Page = self.preparePageToShow( true )
        return pageToShow
    }
    
    func goToPage( pageNum:Int ) -> Page
    {
        self.currentPieceStory = 0
        self.currentPage = 0
        
        //var items:Int = self.story["story"]![self.currentPieceStory]["texts"]!!.count
        var itemsOld:Int = 0
        var foundPage:Bool = false
        while( !foundPage )
        {
            let items:Int = self.story["story"]![self.currentPieceStory]["texts"]!!.count
            if( pageNum <= (items + itemsOld) )
            {
                self.currentPage = pageNum - itemsOld
                foundPage = true
            }
            else
            {
                itemsOld += items
                self.currentPieceStory += 1
            }
        }
        
        let page:Page = self.preparePageToShow( true )
        
        self.checkState()
        
        return page
    }
    
    
    func prevPage() -> Page
    {
        self.currentPage -= 1
        let page:Page = self.checkChangeImage()
        
        self.checkState()

        return page
    }
    
    func nextPage() -> Page
    {
        self.currentPage += 1
        let page:Page = self.checkChangeImage()
        
        self.checkState()
        
        return page
    }
    
    func release()
    {
        //Images.
        for pair in self.imagesPath
        {
            self.imagesPath.removeValueForKey( pair.0 )
        }
        
        //Story
        self.story.removeValueForKey( "story" )
        
        //Story language
        for pair in self.storyLanguage
        {
            self.storyLanguage.removeValueForKey( pair.0 )
        }
    }
    
    func isAtFirstPage() -> Bool
    {
        return self.state == .FrontCover
    }
    
    func isAtLastPage() -> Bool
    {
        return self.state == .BackCover
    }
    
    
    /*=============================================================*/
    /*                       Private Section                       */
    /*=============================================================*/
    private func loadStory()
    {
        let file = NSBundle.mainBundle().pathForResource("story", ofType: "json", inDirectory: "Res/tales/\(self.id)")
        do
        {
            let fileContent = try NSString(contentsOfFile: file!, encoding: NSUTF8StringEncoding )
            //print("Contenido story...\n\(fileContent)")
            
            let fc = fileContent as String!
            self.prepareTexts( fc, type:1 )
            
        }
        catch let error as NSError
        {
            print( "Failed to load: \(error.localizedDescription)" )
        }
    }
    
    private func loadStoryLanguage()
    {
        let file = NSBundle.mainBundle().pathForResource("\(self.langId)", ofType: "json", inDirectory: "Res/tales/\(self.id)/languages")
        do
        {
            let fileContent = try NSString(contentsOfFile: file!, encoding: NSUTF8StringEncoding )
            //print("Contenido storyLangs...\n\(fileContent)")
            
            let fc = fileContent as String!
            self.prepareTexts( fc, type:2 )

        }
        catch let error as NSError
        {
            print( "Failed to load: \(error.localizedDescription)" )
        }
    }
    
    private func prepareTexts(text:String, type:Int)
    {
        switch( type )
        {
            case 1:
                self.story = self.getJSON( text )! as! [String : AnyObject]
                break
            
            case 2:
                self.storyLanguage = self.getJSON( text )! as! [String : String]
                break
            
            default:
                break
        }
    }
    
    private func loadImages()
    {
        let resolutionPath = (self.isHD) ? "/" : "/sd/"
        let story = self.story["story"] as! [AnyObject]
        
        for itemStory in story
        {
            let imgId:String = itemStory["image"]!!.description
            
            let imagePath = NSBundle.mainBundle().pathForResource( "esc\(imgId)", ofType: "jpg", inDirectory: "Res/tales/\(self.id)/images\(resolutionPath)" )
            self.imagesPath[imgId] = imagePath!
        }
    }
    
    private func getJSON(stringJSON:String)->AnyObject?
    {
        let json:AnyObject?
        
        do
        {
            let data = stringJSON.dataUsingEncoding( NSUTF8StringEncoding, allowLossyConversion: false )
            json = try NSJSONSerialization.JSONObjectWithData( data!, options: .AllowFragments )
        }
        catch let error as NSError
        {
            print( "Failed to load: \(error.localizedDescription)" )
            json = nil
        }
        
        return json
    }
    
    private func checkChangeImage() -> Page
    {
        let MAX_PIECE_STORY = (self.story["story"]as![AnyObject]).count - 1
        var changeImage:Bool = false
        self.state = .Page
        
        if( self.currentPage >= self.story["story"]![self.currentPieceStory]["texts"]!!.count )
        {
            //self.currentPieceStory = (self.currentPieceStory < MAX_PIECE_STORY) ? self.currentPieceStory + 1 : 0
            if( self.currentPieceStory < MAX_PIECE_STORY )
            {
                self.currentPieceStory += 1
                self.currentPage = 0
                changeImage = true
            }
            /*
            else
            {
                self.currentPieceStory = MAX_PIECE_STORY
                self.currentPage--
                self.state = .BackCover
            }
            */
        }
        else if( self.currentPage < 0)
        {
            //self.currentPieceStory = (self.currentPieceStory <= 0) ? MAX_PIECE_STORY : self.currentPieceStory - 1
            if( self.currentPieceStory > 0 )
            {
                self.currentPieceStory -= 1
                self.currentPage = self.story["story"]![self.currentPieceStory]["texts"]!!.count - 1
                changeImage = true
            }
            /*
            else
            {
                self.currentPieceStory = 0
                self.currentPage = 0
                self.state = .FrontCover
            }
            */
        }
        
        let pageToShow:Page = self.preparePageToShow( changeImage )
        return pageToShow
    }
    
    private func preparePageToShow(changeImage:Bool) -> Page
    {
        //let textId = self.story["story"]![self.currentPieceStory]["texts"]!![self.currentPage] as! String
        let tmp = self.story["story"]![self.currentPieceStory]["texts"] as! [String]
        let textId = tmp[self.currentPage] 
        let text = self.storyLanguage[textId]!
        let imgPos = self.story["story"]![self.currentPieceStory]["image"] as! String
        let imagePath = self.imagesPath[ imgPos ]
        
        let pageToShow:Page = Page( text:text, imagePath:imagePath!, changeImage: changeImage )
        
        return pageToShow
    }
    
    private func checkState()
    {
        let MAX_PIECE_STORY = (self.story["story"]as![AnyObject]).count - 1
        
        if( self.currentPage == 0 && self.currentPieceStory == 0 )
        {
            self.state = .FrontCover
        }
        else if( self.currentPage >= self.story["story"]![self.currentPieceStory]["texts"]!!.count - 1 && self.currentPieceStory == MAX_PIECE_STORY)
        {
            self.state = .BackCover
        }
    }
    
    
    private var id:String
    private var langId:String
    private var isHD:Bool
    private var story:[String:AnyObject]
    private var storyLanguage:[String:String]
    private var imagesPath:[String:String]
    private var currentPieceStory:Int
    private var currentPage:Int
    private var state:PageType
}

File: android/LiveTales/app/src/main/java/com/zed/livetales/models/book/Book.java
package com.zed.livetales.models.book;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import com.zed.livetales.util.Utils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by jemalpartida on 25/11/2016.
 */

public class Book
{
    public Book(Context context, String id, String langId, boolean isHD)
    {
        this.context = context;
        this.id = id;
        this.langId = langId;
        this.isHD = isHD;
        this.story = null;
        this.storyLanguage = null;
        this.imagesPath = new HashMap<>();
        this.currentPieceStory = 0;
        this.currentPage = 0;
        this.state = PageType.FrontCover;
    }

    public String getId()
    {
        return this.id;
    }

    public Page load()
    {
        this.loadStory();
        this.loadStoryLanguage();
        this.loadImages();

        this.state = PageType.FrontCover;

        Page pageToShow = this.preparePageToShow( true );
        return pageToShow;
    }

    public Page changeLanguage(String langId)
    {
        this.langId = langId;
        this.loadStoryLanguage();

        Page pageToShow = this.preparePageToShow( true );
        return pageToShow;
    }

    public Page goBegin()
    {
        this.currentPieceStory = 0;
        this.currentPage = 0;
        this.state = PageType.FrontCover;

        Page pageToShow = this.preparePageToShow( true );
        return pageToShow;
    }

    public Page goToPage( int pageNum )
    {
        this.currentPieceStory = 0;
        this.currentPage = 0;

        int itemsOld = 0;
        boolean foundPage= false;
        while( !foundPage )
        {
            //int items = this.story["story"]![this.currentPieceStory]["texts"]!!.count

            try {
                int items = this.story.getJSONArray("story").getJSONObject(this.currentPieceStory).getJSONArray("texts").length();

                if (pageNum <= (items + itemsOld)) {
                    this.currentPage = pageNum - itemsOld;
                    foundPage = true;
                } else {
                    itemsOld += items;
                    this.currentPieceStory += 1;
                }
            }
            catch( JSONException jsone )
            {
                Log.e( "TalesLive", "Error getting texts size");
            }
        }

        Page page = this.preparePageToShow( true );

        this.checkState();

        return page;
    }

    public Page prevPage()
    {
        this.currentPage -= 1;
        Page page = this.checkChangeImage();

        this.checkState();

        return page;
    }

    public Page nextPage()
    {
        this.currentPage += 1;
        Page page = this.checkChangeImage();

        this.checkState();

        return page;
    }

    public void release()
    {
        //Images.
        /*
        for pair in this.imagesPath
        {
            this.imagesPath.removeValueForKey( pair.0 )
        }*/
        this.imagesPath.clear();


        //Story
        //this.story.removeValueForKey( "story" )
        this.story = null;
        //Story language
        /*
        for pair in this.storyLanguage
        {
            this.storyLanguage.removeValueForKey( pair.0 )
        }
        */
        this.storyLanguage = null;
    }

    public boolean isAtFirstPage()
    {
        return this.state == PageType.FrontCover;
    }

    public boolean isAtLastPage()
    {
        return this.state == PageType.BackCover;
    }


    /*============================================================================================*/
    /*                                      Private Section                                       */
    /*============================================================================================*/
    private void loadStory()
    {
        try
        {
            AssetManager assetMgr = context.getAssets();
            InputStream is = assetMgr.open( "tales/" + this.id + "/story.json" );
            this.story = Utils.loadJSON( is );
        }
        catch( IOException ioe )
        {
            Log.e( "TalesLive", "Error reading story.json!" );
        }
        catch( JSONException jsone )
        {
            Log.e( "TalesLive", "Error process story.json!" );
        }
    }

    private void loadStoryLanguage()
    {
        try
        {
            AssetManager assetMgr = context.getAssets();
            InputStream is = assetMgr.open( "tales/" + this.id + "/languages/" + this.langId );
            this.storyLanguage = Utils.loadJSON(is);
        }
        catch( IOException ioe )
        {
            Log.e( "TalesLive", "Error reading story.json!" );
        }
        catch( JSONException jsone )
        {
            Log.e( "TalesLive", "Error process story.json!" );
        }
    }

    private void loadImages()
    {
        String resolutionPath = ( this.isHD ) ? "/" : "/sd/";
        try
        {
            JSONArray story = this.story.getJSONArray( "story" );
            for(int itemPos = 0; itemPos < story.length(); itemPos++)
            {
                JSONObject itemStory = story.getJSONObject( itemPos );
                String imgId = itemStory.getString("image");
                String imagePath = "tales/" + this.id + "/images" + resolutionPath + "esc" + imgId + ".jpg";
                this.imagesPath.put( imgId, imagePath );
            }
        }
        catch( JSONException jsone )
        {

        }
    }

    private Page checkChangeImage()
    {
        try
        {
            int MAX_PIECE_STORY = this.story.getJSONArray( "story" ).length();
            boolean changeImage = false;
            this.state = PageType.Page;

            int count = this.story.getJSONArray("story").getJSONObject(this.currentPieceStory).getJSONArray("texts").length();
            if( this.currentPage >= count )
            {
                if( this.currentPieceStory < MAX_PIECE_STORY )
                {
                    this.currentPieceStory += 1;
                    this.currentPage = 0;
                    changeImage = true;
                }
            }
            else if( this.currentPage < 0 )
            {
                if( this.currentPieceStory > 0 )
                {
                    this.currentPieceStory -= 1;
                    this.currentPage = count - 1;
                    changeImage = true;
                }
            }

            Page pageToShow = this.preparePageToShow( changeImage );
            return pageToShow;
        }
        catch( JSONException jsone )
        {
            Log.e( "LiveTales", "Error accediendo texts!" );
            return null;
        }

    }

    private Page preparePageToShow(boolean changeImage)
    {
        try
        {
            JSONArray tmp = this.story.getJSONArray("story").getJSONObject(this.currentPieceStory).getJSONArray("texts");
            String textId = tmp.getString(this.currentPage);
            String text = this.storyLanguage.getString(textId);
            String imgPos = this.story.getJSONArray("story").getJSONObject(this.currentPieceStory).getString("image");
            String imagePath = this.imagesPath.get(imgPos);

            Page pageToShow = new Page(text, imagePath, changeImage);
            return pageToShow;
        }
        catch(JSONException jsone )
        {
            Log.e("","");
            return null;
        }
    }

    private void checkState()
    {
        try
        {
            int MAX_PIECE_STORY = this.story.getJSONArray("story").length() - 1;

            if( this.currentPage == 0 && this.currentPieceStory == 0 )
            {
                this.state = PageType.FrontCover;
            }
            else if( this.currentPage >= this.story.getJSONArray("story").getJSONObject(this.currentPieceStory).getJSONArray("texts").length() - 1 && this.currentPieceStory == MAX_PIECE_STORY)
            {
                this.state = PageType.BackCover;
            }
        }
        catch( JSONException jsone )
        {
            Log.e( "", "" );
        }
    }


    private Context context;
    private String id;
    private String langId;
    private boolean isHD;
    private JSONObject story;
    private JSONObject storyLanguage;
    private Map<String,String> imagesPath;
    private int currentPieceStory;
    private int currentPage;
    private PageType state;
}
